engine/rpcserver.go

```diff
@@ -4749,3 +4750,180 @@ func (s *RPCServer) GetTechnicalAnalysis(ctx context.Context, r *gctrpc.GetTechn
+		}
+
+		if request.TakeFeeRate.LessThanOrEqual(decimal.Zero) {
+			return nil, fmt.Errorf("%w for offline calculations", common.ErrCannotCalculateOffline)
+		}
+		if len(r.Rates) == 0 {
+			return nil, fmt.Errorf("%w for offline calculations", common.ErrCannotCalculateOffline)
+		}
+		for i := range r.Rates {
+			var offlineRate margin.Rate
+			offlineRate.Time, err = time.Parse(common.SimpleTimeFormat, r.Rates[i].Time)
+			if err != nil {
+				return nil, err
+			}
+
+			offlineRate.HourlyRate, err = decimal.NewFromString(r.Rates[i].HourlyRate)
+			if err != nil {
+				return nil, err
+			}
+
+			if r.Rates[i].BorrowCost != nil {
+				offlineRate.BorrowCost.Size, err = decimal.NewFromString(r.Rates[i].BorrowCost.Size)
+				if err != nil {
+					return nil, err
+				}
+			}
+			if r.Rates[i].LendingPayment != nil {
+				offlineRate.LendingPayment.Size, err = decimal.NewFromString(r.Rates[i].LendingPayment.Size)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			request.Rates = append(request.Rates, offlineRate)
+		}
+	}
+
+	lendingResp, err := exch.GetMarginRatesHistory(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	if len(lendingResp.Rates) == 0 {
+		return nil, order.ErrNoRates
+	}
+	resp := &gctrpc.GetMarginRatesHistoryResponse{
+		LatestRate: &gctrpc.MarginRate{
+			Time:             lendingResp.Rates[len(lendingResp.Rates)-1].Time.Format(common.SimpleTimeFormatWithTimezone),
+			HourlyRate:       lendingResp.Rates[len(lendingResp.Rates)-1].HourlyRate.String(),
+			YearlyRate:       lendingResp.Rates[len(lendingResp.Rates)-1].YearlyRate.String(),
+			MarketBorrowSize: lendingResp.Rates[len(lendingResp.Rates)-1].MarketBorrowSize.String(),
+		},
+		TotalRates: int64(len(lendingResp.Rates)),
+	}
+	if r.GetBorrowRates {
+		resp.LatestRate.HourlyBorrowRate = lendingResp.Rates[len(lendingResp.Rates)-1].HourlyBorrowRate.String()
+		resp.LatestRate.YearlyBorrowRate = lendingResp.Rates[len(lendingResp.Rates)-1].YearlyBorrowRate.String()
+	}
+	if r.GetBorrowRates || r.GetLendingPayments {
+		resp.TakerFeeRate = lendingResp.TakerFeeRate.String()
+	}
+	if r.GetLendingPayments {
+		resp.SumLendingPayments = lendingResp.SumLendingPayments.String()
+		resp.AvgLendingSize = lendingResp.AverageLendingSize.String()
+	}
+	if r.GetBorrowCosts {
+		resp.SumBorrowCosts = lendingResp.SumBorrowCosts.String()
+		resp.AvgBorrowSize = lendingResp.AverageBorrowSize.String()
+	}
+	if r.GetPredictedRate {
+		resp.PredictedRate = &gctrpc.MarginRate{
+			Time:       lendingResp.PredictedRate.Time.Format(common.SimpleTimeFormatWithTimezone),
+			HourlyRate: lendingResp.PredictedRate.HourlyRate.String(),
+			YearlyRate: lendingResp.PredictedRate.YearlyRate.String(),
+		}
+		if r.GetBorrowRates {
+			resp.PredictedRate.HourlyBorrowRate = lendingResp.PredictedRate.HourlyBorrowRate.String()
+			resp.PredictedRate.YearlyBorrowRate = lendingResp.PredictedRate.YearlyBorrowRate.String()
+		}
+	}
+	if r.IncludeAllRates {
+		for i := range lendingResp.Rates {
```

Comment:
Can add a `resp.Rates = make([]*gctrpc.MarginRate, len(lendingResp.Rates))` here and then `resp.Rates[i] = rate` at the end of the for loop